cpp/src/merge/merge.cu

```diff
@@ -13,417 +13,509 @@
+   }
+ 
+ public:
+   merge_queue_item(table_view const& view, table_ptr_type&& table):
+   view_(view), 
+   table_(std::move(table)) {
+     set_priority();
+   }
+
+   explicit merge_queue_item(std::priority_queue<merge_queue_item>& queue){
+     auto& item = const_cast<merge_queue_item&>(queue.top());
+     view_ = item.view_;
+     // Safe as priority_ is not affected by the move
+     table_ = std::move(item.table_);
+     set_priority();
+
+     queue.pop();
+   }
+
+   auto& view() const { return view_; }
+   auto& table() { return table_; }
+ 
+   bool operator<(merge_queue_item const& other) const {
+       return priority_ < other.priority_;
+   }
+ };
+ 
+ table_ptr_type merge(std::vector<table_view> const& tables_to_merge,
+                      std::vector<cudf::size_type> const& key_cols,
+                      std::vector<cudf::order> const& column_order,
+                      std::vector<cudf::null_order> const& null_precedence,
+                      rmm::mr::device_memory_resource* mr,
+                      cudaStream_t stream = 0) {
+     // TODO add test case
+     if (tables_to_merge.empty()) {
+       return std::make_unique<cudf::experimental::table>();
+     }
+ 
+     auto const& first_table = tables_to_merge.front();
+     const auto n_cols = first_table.num_columns();
+ 
+     for (auto const& table: tables_to_merge) {
+       CUDF_EXPECTS(n_cols == table.num_columns(),
+                    "Mismatched number of columns");
+       CUDF_EXPECTS(cudf::have_same_types(first_table, table),
+                    "Mismatched column types");
+     }
+ 
+     CUDF_EXPECTS(!key_cols.empty(),
+                  "Empty key_cols");
+     CUDF_EXPECTS(key_cols.size() <= static_cast<size_t>(n_cols), 
+                  "Too many values in key_cols");
+ 
+     CUDF_EXPECTS(key_cols.size() == column_order.size(),
+                  "Mismatched size between key_cols and column_order");
+     CUDF_EXPECTS(column_order.size() <= static_cast<size_t>(n_cols), 
+                  "Too many values in column_order");
+ 
+     if (tables_to_merge.size() == 1 || n_cols == 0) {
+       // TODO add test case
+       return std::make_unique<cudf::experimental::table>(first_table);
+     }
+ 
+     // A queue of (table view, table) pairs
+     std::priority_queue<merge_queue_item> merge_queue;
+     // The table pointer is null if we do not own the table (input tables)
+     std::for_each(tables_to_merge.begin(), 
+                   tables_to_merge.end(), 
+                   [&](auto const& table) { 
+                     if (table.num_rows() > 0)
+                       merge_queue.emplace(table, table_ptr_type()); 
+                   });
+     // If there is only one table in the queue, that's the final result
+     while (merge_queue.size() > 1) {
+       // TODO explain the traversal logic
+       
+       // Deallocated at the end of the block
+       const auto left_table = merge_queue_item(merge_queue);
```

Comment:
It's not obvious that this ctor would modify the `merge_queue`. I think the intent would be clearer to do something like:
```c++
auto const left_table = merge_queue.top();
merge_queue.pop();

auto const right_table = merge_queue.top();
merge_queue.pop();
```

I've always thought it was unfortunate that there wasn't a member of `priority_queue` that returns the first element and pops it at the same time.

Reply:
@jrhemstad  Unfortunately this does not compile. The first line requires a copy constructor.

Reply:
`auto const left_table = std::move(merge_queue.top());`?

Reply:
Also does not work without the const_cast because it returns a const& which cannot bind to &&.

Reply:
Yuck, that's annoying

Reply:
Okay, how about this:
```c++
template <typename T>
T top_and_pop(std::priority_queue<T>& q){
  auto moved = std::move(const_cast<T&>(q.top())));
  q.pop();
  return moved;
}
```